### ISSUE TITLE
daily/weekly/monthly tasks run in parallel, causing slowdown.

### ISSUE THREAD

**Issue by ian-kelling** [kdave/btrfsmaintenance#15]
For example, in the default configuration, weekly btrfs balance will run at the same time as monthly btrfs scrub if the balance takes longer than 7 minutes to run on debian. This should be avoided, or at least have an option to avoid it. On debian, with anacron or cron, daily/weekly/monthly start 5-22 after each other. Each of daily/weekly/monthly will run it's own tasks in sequence, not parallel.

**Comment by kdave** [comment]
That could be a big hit. Can we return a value from the script to tell cron we'd like to run again later? Also, the systemd timers could start the services so this would need to be reflected in some way. But let's address cron first.

**Comment by ian-kelling** [comment]
If these were native systemd timers, I'd think you'd probably solve all of this with a simple config option in each timer file: conflicts=other_btrfs_maintenance_job.timer. I'm going to try it out on my system.

There is no way to tell cron we want to run again later through return code. In the cron realm, I think the right way would be to have a lockfile and sleep in a loop if the lockfile exists at the beginning of the scripts. something like,

```
while [ -e /var/lock/btrfsmaintenance ]; do sleep 60; done
echo $$ > /var/lock/btrfsmaintenance
trap 'rm -f /var/lock/btrfsmaintenance; exit $?' INT TERM EXIT
```


And I'd probably set a timeout for the while loop to be safe, like have a default timeout of half a day and increase it if we can tell we are running from cron.weekly or cron.monthly, to half a week or half a month. And a separate issue which this exacerbates, btrfs scrub can take over a day and is already blocking other cron scripts alphabetically after it which are in the same cron directory, so I would wrap it in another script and call it with "script &" to put  in the background, then it won't block other cronjobs from running. balance can take a while too, so I'd do it for both.

**Comment by ian-kelling** [comment]
Actually, I'm not going to test the systemd timers, I just put balance and scrub into the same cron directory for my own system. Everything else I said still stands.

**Comment by kdave** [comment]
With custom locking and waiting, the jobs will become unpredictable and possibly hard for the administrator to manage once the cron/timer fires. But as the jobs are periodical, I think we can afford to skip a round, if there's a conflicting job running. This would  need to define some priorities and give some recommended order of execution among the cron scripts.

Outline of the simple approach:
* scrub: runs less often, is not absolutely necessary, delaying is not crucial
* balance: as it compacts the space, it affects daily use, dealy is undesirable but skipping one round is acceptable (IMO)

The logic: when a job sees a conflicting one, it ends. This could lead to a job being skipped for a long time, so cron should be configured to run the jobs with the longest period first (a minute-range offset should suffice).
This is however beyond what the scope of the provided cron directories like daily/weekly/monthly. As we have all the information in the config, we can add the delays aftificially (configurable).
Ideally we could also look into a log of "when the job was last run", if it's less than the period, then don't delay at all, but in case more than one period fires, own delays would apply.

**Comment by ian-kelling** [comment]
Sounds reasonable.

**Comment by barak** [comment]
In Debian, this could be dealt with by adding lines to each script that grab/release a global btrfsmaintenance  lock before/after doing their business.
```
dotlockfile -r -1 -p -l /var/run/btrfsmaintenance.pid
echo DO BUSINESS HERE
dotlockfile -u /var/run/btrfsmaintenance.pid
```
The command is in package `liblockfile-bin`. The `-r -1` says to wait and retry grabbing the lock forever; a timeout of some length would be reasonable too, or maybe make it configurable.

**Comment by Gatak** [comment]
I'd like some more logic around this topic. For example, running balance, scrub or snapshots while a btrfs send|receive is happening is not a good thing. I've experienced many difficult locks like this.

Perhaps some better way of detecting if any btrfs if task is running on the device before running the schedules task would be good.

**Comment by kdave** [comment]
The file locking exclusion has been implemented, that's better than nothing. The fine grained configuration can be achieved by the calendar event specification in the timer units. This could simulate the ordering. With this I'm not sure if adding the logic into the scripts won't make it less predictable.

**Comment by barak** [comment]
Nice!

**Comment by kdave** [comment]
Fixed in 0.5: systemd.timer calendar format can be used set any time, there's the task exclusion in place too. Closing.
